src/lib.rs

```diff
@@ -33,48 +30,68 @@ pub struct UrlEncodedQuery;
 /// Use it like this: `req.get_ref::<UrlEncodedBody>()`
 pub struct UrlEncodedBody;
 
+#[derive(Clone, Debug, PartialEq)]
+pub enum ErrorType{
+    EncodingError(Utf8Error),
+    EmptyQuery
+}
+
 /// Hashmap mapping strings to vectors of strings.
 pub type QueryMap = HashMap<String, Vec<String>>;
 
-impl Assoc<QueryMap> for UrlEncodedQuery {}
-impl Assoc<QueryMap> for UrlEncodedBody {}
+impl Pluggable for UrlEncodedQuery {}
+
+impl Pluggable for UrlEncodedBody {}
+
+impl Key for UrlEncodedQuery {
+    type Value = QueryMap;
+}
 
-impl PluginFor<Request, QueryMap> for UrlEncodedQuery {
-    fn eval(req: &mut Request, _: Phantom<UrlEncodedQuery>) -> Option<QueryMap> {
+impl Key for UrlEncodedBody {
+    type Value = QueryMap;
+}
+
+impl<'a> Plugin<Request<'a>> for UrlEncodedQuery {
+    type Error = ErrorType;
+    fn eval(req: &mut Request) -> Result<QueryMap, ErrorType> {
         match req.url.query {
-            Some(ref query) => create_param_hashmap(query.as_slice()),
-            None => None
+            Some(ref query) => create_param_hashmap(&query[]),
+            None => Err(ErrorType::EmptyQuery)
         }
     }
 }
 
-
-impl PluginFor<Request, HashMap<String, Vec<String>>> for UrlEncodedBody {
-    fn eval(req: &mut Request, _: Phantom<UrlEncodedBody>) -> Option<QueryMap> {
-        str::from_utf8(req.body.as_slice()).and_then(create_param_hashmap)
+impl<'a> Plugin<Request<'a>> for UrlEncodedBody {
+    type Error = ErrorType;
+    fn eval(req: &mut Request) -> Result<QueryMap, ErrorType> {
+        str::from_utf8(&req.body.read_to_end().unwrap())
```

Comment:
Sorry, you'll have to be more explicit.  What (and where) is the body reader plugin?